lib/CommandClient.php

```diff
@@ -26,22 +32,82 @@ public static function socketPath(string $config) {
     private function send($message): Promise {
         return call(function () use ($message) {
             if (!$this->socket) {
-                $this->socket = yield from $this->establish();
+                yield from $this->establish();
             }
 
-            $message = \json_encode($message);
-            $message = \pack("N", \strlen($message)) . $message;
-            return yield $this->socket->write($message);
+            try {
+                $message = \json_encode($message) . "\n";
+                yield $this->socket->write($message);
+            } catch (ClosedException $e) {
+                yield from $this->establish();
+                yield $this->socket->write($message);
+            }
+
+            $promise = ($this->readDeferreds[] = new Deferred)->promise();
+
+            // we can ONLY resolve the socketDeferred AFTER the first write, otherwise order of send()'s is not respected
+            if ($this->socketDeferred) {
+                $deferred = $this->socketDeferred;
+                $this->socketDeferred = null;
+                $deferred->resolve();
+            }
+
+            return $promise;
         });
     }
 
+    private static function read(&$socket, &$readDeferreds) {
+        $parser = new \Amp\Parser\Parser((function () use (&$readDeferreds) {
+            do {
+                $message = @\json_decode(yield "\n", true);
+                if (isset($message["exception"])) {
+                    \current($readDeferreds)->fail(new \RuntimeException($message["exception"]));
+                } else {
+                    \current($readDeferreds)->resolve($message);
+                }
+                unset($readDeferreds[\key($readDeferreds)]);
+            } while (true);
+        })());
+
+        while (($chunk = yield $socket->read()) !== null) {
+            $parser->push($chunk);
+        }
+
+        $socket = null;
+        $deferreds = $readDeferreds;
+        $readDeferreds = [];
+        foreach ($deferreds as $deferred) {
+            $deferred->fail(new ClosedException("Connection went away ..."));
+        }
+    }
```

Comment:
Assuming that you will continue to maintain amp forever. Assuming that we don't release dependency upgrades causing small maintenance.

… I'm just not going to buy these assumptions, for any package within amp. I'd like to be consequent here and just won't do that.

If we make that class part of ByteStream package for example, I'm totally fine with using it though.